Ndev 2299 debug trace transaction call tracer

romanova-natasha opened the pull request.

romanova-natasha committed:
add cases for debug_trace_transaction with callTracer

add cases for debug_trace_transaction with callTracer

romanova-natasha committed:
add cases calling contract from another contract with different reverts

add cases calling contract from another contract with different reverts

romanova-natasha committed:
add cases with logs in different response level, add deep diff check of response, add case: check trace of precompiled

add cases with logs in different response level, add deep diff check of response, add case: check trace of precompiled

kristinaNikolaevaa commented on contracts/common/tracer/ContractCaller.sol:
```diff
@@ -0,0 +1,156 @@
+    function lowLevelCallContract(
+        address _contractCallee
+    ) public returns (bool) {
+        bytes memory payload = abi.encodeWithSignature("returnNumber()", 1);
```
Are you sure that it works? I think there should be types in signature, like
"returnNumber(uint256)"

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -355,6 +355,37 @@ def event_caller_contract(web3_client, accounts) -> tp.Any:
     yield event_caller
 
+@pytest.fixture(scope="class")
+def caller_contract(web3_client, accounts) -> tp.Any:
```
let's rename it please or put to tracer conftest, because we have a lot of caller contracts

romanova-natasha replied:
done

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -355,6 +355,37 @@ def event_caller_contract(web3_client, accounts) -> tp.Any:
+    contract, _ = web3_client.deploy_and_get_contract(
+        "common/tracer/ContractCaller", "0.8.15", account=accounts[0]
+    )
+    yield contract, address
```
So, you are return caller contract and  address of callee contract, maybe it's beter to create separeted fixures? It's difficult to read

romanova-natasha replied:
done

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+    
+    def assert_response_contains_expected(self, expected_response, response):
+        diff = DeepDiff(expected_response, response["result"])
+        assert ("dictionary_item_added" and "dictionary_item_removed") not in diff
```
maybe it's better to check that the diff is empty? or to add 'values_changed' to assert? Or you dont want to check values here at all?

romanova-natasha replied:
values_changed and dictionary_item_removed are added, so we can check if expected response is a subset of actual response and all identical fields match

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+            method="debug_traceTransaction", params=[receipt["transactionHash"].hex(), tracer_params]
+        )
+        expected_response = self.fill_expected_response(instruction_tx, receipt, calls_value="0x0", calls_type="CREATE2")
+        result_diff = DeepDiff(expected_response, response["result"])
```
assert_response_contains_expected() ?

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+
+        tx_obj, _, receipt = call_storage(sender_account, storage_contract, store_value, "blockNumber", self.web3_client)
+        
+        wait_condition(
```
Let's move this debug_traceTransaction call with waiting for result to tracer_api class? It is used at the same way everywhere

romanova-natasha replied:
tracer_api is just a JsonRpcSession fixture
we use tracer_api fixture for making requests just like json_rpc_client fixture for proxy rpc tests
Could you please explain in other words what do you mean here and what do you want to change?

kristinaNikolaevaa replied:
I ment You can create the class Tracer_api_client as a child of JsonRPCSession and move there 
```
        wait_condition(
            lambda: self.tracer_api.send_rpc(
                method="debug_traceTransaction", params=[receipt["transactionHash"].hex(), tracer_params]
            )["result"]
            is not None,
            timeout_sec=120,
        )

        return self.tracer_api.send_rpc(
            method="debug_traceTransaction", params=[receipt["transactionHash"].hex(), tracer_params]
        )
        
        
        do not repeat every time
```

kristinaNikolaevaa replied:
Because it can be used in deferent classes for tests

romanova-natasha replied:
Agree, I added a class tracer_client

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+        expected_response = self.fill_expected_response(instruction_tx, receipt, revert=True, revert_reason=reason, calls_value="0x0")
+        self.assert_response_contains_expected(expected_response, response)
+    
+    def test_callTracer_call_contract_from_other_contract_revert_with_assert(self, caller_contract):
```
What are you checking in these tests?Looks like they fails on call because of Insufficient balance for transfer, not because of "assert(false)" for example. Earlier and at the same way, on inner call

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+    
+    def test_callTracer_call_contract_from_other_contract_revert_with_require(self, caller_contract):
+        sender_account = self.accounts[0]
+        contractTwo = caller_contract[0]
```
let's use snake_case for names as everywhery

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on contracts/common/tracer/ContractCaller.sol:
```diff
@@ -0,0 +1,156 @@
+        require(success);
+    }
+
+    function setParamWithCallcode(
```
I can't find where you use this function

romanova-natasha replied:
deleted

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+        )
+
+        expected_response["from"] = instruction_tx["from"].lower()
+        expected_response["to"] = instruction_tx["to"].lower()
```
Isn't fill_expected_response appropriate for some tests?

romanova-natasha replied:
I fixed this, I added fill_expected_response to all cases except one test_callTracer_type_create (we do not have tx_instruction for this case and I check response as it is)

romanova-natasha committed:
small refactoring of the cases, rm code from contracts which is not used in these tests

small refactoring of the cases, rm code from contracts which is not used in these tests

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+@allure.feature("Tracer API")
+@allure.story("Tracer API RPC calls debug method trace_transaction callTracer check")
+@pytest.mark.usefixtures("accounts", "web3_client", "tracer_api")
+class TestDebugTraceTransactionCallTracer:
```
Do you have tests for cases:
1) creating contracts in a constructor and emitting events in constructor
2) delegatecall + logs 
3) different value for onlyTopCall
?

romanova-natasha replied:
1) case is added
2) we have call + logs and static call + logs, I am not sure we have to check all combinations. But if you think we should I'll fix it. Just let me know your decision
3) with onlyTopCall we have an issue: it works strange to me but I need to discuss it with Silviu. He is out before May 8 so I just created a bug NDEV-2959 and added a link to test case with skip mark for now. We will manage this case later with Silviu, Is it ok for you?

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+                    expected_response["calls"].append({
+                    "from": address_to,
+                    "gasUsed":"0x0",
+                    "gas": "0x0",
```
Is it always =0 for inner calls? Is it the same in ethereum?

romanova-natasha replied:
it is a known issue I added a comment to the code, idk when and how it will be resolved. The only doc we can refer here is our private notion page so I have decided just add the comment for now

kristinaNikolaevaa commented on integration/tests/tracer/test_tracer_debug_trace_transactions_callTracer.py:
```diff
@@ -0,0 +1,501 @@
+
+        if logs:
+            for log in receipt["logs"]:
+                if log["logIndex"] == 0:
```
So, you are checking only 2 log events (1 external and 1 inner) 
Should we have a test to check case when there are more than 1 event ?

romanova-natasha replied:
another one case is added: test_callTracer_call_contract_with_event_from_other_one_with_two_events

romanova-natasha committed:
fix revert calls: rm payable, fix expected and actual response comparison, refactor expected response filling, add case with create type check

fix revert calls: rm payable, fix expected and actual response comparison, refactor expected response filling, add case with create type check

romanova-natasha committed:
add: sort calls

add sort calls

romanova-natasha committed:
add OnlyTopCall case with mark skip, add case with two events emitted from one contract

add OnlyTopCall case with mark skip, add case with two events emitted from one contract

romanova-natasha committed:
add case with new contact and event emition from constructor of other contract and check response tree

add case with new contact and event emition from constructor of other contract and check response tree

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -356,6 +356,41 @@ def event_caller_contract(web3_client, accounts) -> tp.Any:
+
+
+@pytest.fixture(scope="class")
+def tracer_calle_contract_address(web3_client, accounts) -> tp.Any:
```
callee?

romanova-natasha replied:
fixed

romanova-natasha committed:
add tracer_client class for working with tracer requests, fix typo

add tracer_client class for working with tracer requests, fix typo

romanova-natasha committed:
Merge branch 'develop' into NDEV-2299-debug-trace-transaction-callTracer
